src/main/java/io/appium/java_client/TouchAction.java

```diff
@@ -177,49 +227,91 @@ public TouchAction tap(int x, int y) {
      * @param x  x offset.
      * @param y  y offset.
      * @return this TouchAction, for chaining.
+     * @deprecated use {@link #tap(PositionOffsetOption)} instead.
      */
-    public TouchAction tap(WebElement el, int x, int y) {
-        ActionParameter action = new ActionParameter("tap", (HasIdentity) el);
-        action.addParameter("x", x);
-        action.addParameter("y", y);
-        parameterBuilder.add(action);
-        return this;
+    @Deprecated
+    public T tap(WebElement el, int x, int y) {
+        return  tap(elementOption(el, x, y));
     }
 
     /**
      * A wait action, used as a NOP in multi-chaining.
      *
      * @return this TouchAction, for chaining.
      */
-    public TouchAction waitAction() {
+    public T waitAction() {
         ActionParameter action = new ActionParameter("wait");
         parameterBuilder.add(action);
-        return this;
+        //noinspection unchecked
+        return (T) this;
+    }
+
+    /**
+     * Waits for specified amount of time to pass before continue to next touch action.
+     *
+     * @param waitOptions see {@link WaitOptions}.
+     * @return this TouchAction, for chaining.
+     */
+    public T waitAction(WaitOptions waitOptions) {
+        ActionParameter action = new ActionParameter("wait", waitOptions);
+        parameterBuilder.add(action);
+        //noinspection unchecked
+        return (T) this;
     }
 
     /**
      * Waits for specified amount of time to pass before continue to next touch action.
      *
      * @param duration of the wait action. Minimum time reolution unit is one millisecond.
      * @return this TouchAction, for chaining.
+     * @deprecated use {@link #waitAction(WaitOptions)} instead.
      */
-    public TouchAction waitAction(Duration duration) {
-        ActionParameter action = new ActionParameter("wait");
-        action.addParameter("ms", duration.toMillis());
+    @Deprecated
+    public T waitAction(Duration duration) {
+        ActionParameter action = new ActionParameter("wait",
+                new WaitOptions()
+                        .withDuration(duration));
         parameterBuilder.add(action);
-        return this;
+        //noinspection unchecked
+        return (T) this;
+    }
+
+    /**
+     * Press and hold the at the center of an element until the context menu event has fired.
+     *
+     * @param longPressOptions see {@link LongPressOptions}.
+     * @return this TouchAction, for chaining.
+     */
+    public T longPress(LongPressOptions longPressOptions) {
```

Comment:
do we still need this method?

Reply:
@mykola-mokhnach 
Yes because it has the additional 
```
withDuration
```